pkg/chaosdaemon/iptables_server.go

```diff
@@ -42,144 +42,169 @@ func (s *daemonServer) SetIptablesChains(ctx context.Context, req *pb.IptablesCh
 	nsPath := GetNsPath(pid, bpm.NetNS)
 
 	iptables := buildIptablesClient(ctx, nsPath)
-	err = iptables.initializeEnv()
-	if err != nil {
-		log.Error(err, "error while initializing iptables")
-		return nil, err
-	}
 
-	err = iptables.setIptablesChains(req.Chains)
-	if err != nil {
-		log.Error(err, "error while setting iptables chains")
-		return nil, err
+	for _, chain := range req.Chains {
+		err = iptables.setIptablesChain(chain)
+		if err != nil {
+			log.Error(err, "error while set iptables chains")
+			return nil, err
+		}
 	}
-
 	return &empty.Empty{}, nil
 }
 
-type iptablesClient struct {
-	ctx    context.Context
-	nsPath string
-}
-
-type iptablesChain struct {
-	Name  string
-	Rules []string
-}
-
-func buildIptablesClient(ctx context.Context, nsPath string) iptablesClient {
-	return iptablesClient{
-		ctx,
-		nsPath,
-	}
-}
-
-func (iptables *iptablesClient) setIptablesChains(chains []*pb.Chain) error {
-	for _, chain := range chains {
-		err := iptables.setIptablesChain(chain)
+func (iptables *iptablesClient) setIptablesChain(chain *pb.Chain) error {
+	switch chain.Command {
+	case pb.Chain_NEW:
+		var rule string
+		if chain.Table != "" {
+			rule = fmt.Sprintf("-t %s ", chain.Table)
+		}
+		log.Info("Create New Chain ", chain.ChainName, rule)
+		err := iptables.createNewChain(&iptablesChain{
+			Name: chain.ChainName,
+			Rule: rule,
+		})
+		if err != nil {
+			log.Error(err, "error while create iptables chains")
+			return err
+		}
+	case pb.Chain_ADD:
+		rule, err := parseAddChain(chain)
+		log.Info("Add New rule in ", chain.ChainName, rule)
+		if err != nil {
+			log.Error(err, "error while add iptables chains")
+			return err
+		}
+		if len(chain.Ipsets) > 1 {
+			for _, ipset := range chain.Ipsets {
+				rule = fmt.Sprintf(rule, ipset)
+				err := iptables.ensureRule(&iptablesChain{
+					Name: chain.ChainName,
+				}, rule)
+				if err != nil {
+					log.Error(err, "error while add iptables chains")
+					return err
+				}
+			}
+		} else {
+			err := iptables.ensureRule(&iptablesChain{
+				Name: chain.ChainName,
+			}, rule)
+			if err != nil {
+				log.Error(err, "error while add iptables chains")
+				return err
+			}
+		}
+	case pb.Chain_DELETE:
```

Comment:
The design of `SetIptablesChains` is purely functional, which means it describes all chains (added by chaos-mesh) on a pod, but not an operation on the IPTables.